src/librustc/traits/util.rs

```diff
@@ -120,24 +139,21 @@ impl<'cx, 'gcx, 'tcx> Elaborator<'cx, 'gcx, 'tcx> {
         let tcx = self.visited.tcx;
         match *predicate {
             ty::Predicate::Trait(ref data) => {
-                // Predicates declared on the trait.
+                // Get predicates declared on the trait.
                 let predicates = tcx.super_predicates_of(data.def_id());
 
-                let mut predicates: Vec<_> =
-                    predicates.predicates
-                              .iter()
-                              .map(|(p, _)| p.subst_supertrait(tcx, &data.to_poly_trait_ref()))
-                              .collect();
-
+                let mut predicates: Vec<_> = predicates.predicates
```

Comment:
It's preexisting, but this can be rewritten without an intermediate `Vec`, by filtering after the `map` and directly extending `self.stack`. If printing is desired, storing the size of `self.stack` before extending it should allow you to get a slice of all the new elements.

Reply:
@oli-obk And change the `retain` into a `filter` I suppose? Makes sense; good idea.

Reply:
Ah, the reason was the borrow checker: it needs the collect, else `self` is still borrowed when you try to do `self.stack.extend`...